### PR TITLE
Add some basic CUDA classes

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/foundation/cuda/memory.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.cpp]: `// Interface header.`

File: src/appleseed/foundation/cuda/memory.h [hunk 2]
```diff
@@ -0,0 +1,189 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2018 Esteban Tovagliari, The appleseedhq Organization
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.h]: 2019

File: src/appleseed/foundation/cuda/memory.h [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_FOUNDATION_CUDA_MEMORY_H
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.h]: `#pragma once`

File: src/appleseed/foundation/cuda/memory.h [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+    static CUresult free(pointer_type p);
+
+    static CUdeviceptr get_device_ptr(const pointer_type p)
+    {
+        return p;
+    }
+};
+
+
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.h]: Only one blank line ?

File: src/appleseed/foundation/cuda/module.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2018 Esteban Tovagliari, The appleseedhq Organization
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/module.cpp]: 2019

File: src/appleseed/foundation/cuda/module.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
+#include "foundation/cuda/exception.h"
+#include "foundation/cuda/module.h"
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/module.cpp]: `// Interface header.`

**Comment by est77** [comment]
> I'm really interest about what you are doing, could you give me some hint ? Like what is the design you have in mind ?

We don't have a specific plan at the moment. We can probably start doing some planning after the next release. For now, we are merging basic things that are likely to be needed in all cases, like basic memory management code, kernel compilation targets for cmake and CUDA related defines.

File: src/appleseed/foundation/cuda/memory.cpp [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+// Interface header.
+#include "foundation/cuda/memory.h"
+
+namespace foundation
+{
+
+GlobalDeviceMemoryAllocator::pointer_type GlobalDeviceMemoryAllocator::alloc(
+    const size_t        size,
+    const unsigned int  flags)
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.cpp]: `flags` is not used at all

Comment (est77) [reply src/appleseed/foundation/cuda/memory.cpp]: I want all the allocators to have the same interface to easily switch between them.

File: src/appleseed/foundation/cuda/memory.h (hunk 4 above)
Comment (est77) [reply src/appleseed/foundation/cuda/memory.h]: We usually use two spaces between classes

File: src/appleseed/foundation/cuda/module.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+class CUDAModule
+  : NonCopyable
+{
+  public:
+    CUDAModule();
+
+    explicit CUDAModule(const char* filename);
+
+    CUDAModule(CUDAModule&& other);
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/module.h]: Why do you have a double reference ?

File: src/appleseed/foundation/cuda/memory.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+{
+    pointer_type p;
+    check_cuda_result(cuMemHostAlloc(&p, size, flags));
+    return p;
+}
+
+CUresult ZeroCopyDeviceMemoryAllocator::free(pointer_type p)
+{
+    return cuMemFreeHost(p);
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/memory.cpp]: Don't you assert the ptr here ?

File: src/appleseed/foundation/cuda/module.h (hunk 8 above)
Comment (est77) [reply src/appleseed/foundation/cuda/module.h]: It is a move constructor. The class is not copyable, but I still want to store it into containters and things like that.

File: src/appleseed/foundation/cuda/memory.cpp (hunk 9 above)
Comment (est77) [reply src/appleseed/foundation/cuda/memory.cpp]: I think that in the same was as delete or free, freeing a null pointer in CUDA does nothing.

File: src/appleseed/foundation/cuda/memory.h [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+//
+// DeviceBuffer class.
+//
+
+template <typename T, typename DeviceMemoryAllocator = GlobalDeviceMemoryAllocator>
+class DeviceBuffer
+  : NonCopyable
```
Comment (dictoon) [inline src/appleseed/foundation/cuda/memory.h]: `public`

File: src/appleseed/foundation/cuda/memory.h [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+// appleseed.foundation headers.
+#include "foundation/core/concepts/noncopyable.h"
+#include "foundation/cuda/exception.h"
+
+// CUDA headers.
+#include "cuda.h"
+
+// Standard headers.
+#include <cassert>
```
Comment (dictoon) [inline src/appleseed/foundation/cuda/memory.h]: Include `<cstddef>`.

File: src/appleseed/foundation/cuda/memory.h [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    explicit DeviceBuffer(const size_t size, const unsigned int flags = 0)
+      : m_ptr(0)
+      , m_size(0)
+    {
+        reset(size, flags);
+    }
+
+    ~DeviceBuffer()
+    {
+        if (m_ptr)
```
Comment (dictoon) [inline src/appleseed/foundation/cuda/memory.h]: I suspect that this `if` is unnecessary.

File: src/appleseed/foundation/cuda/module.h [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace foundation
+{
+
+//
+// Module class.
+//
+
+class CUDAModule
+  : NonCopyable
```
Comment (dictoon) [inline src/appleseed/foundation/cuda/module.h]: `public`

File: src/appleseed/foundation/cuda/module.cpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+CUDAModule::~CUDAModule()
+{
+    if (m_module)
+        cuModuleUnload(m_module);
+}
+
+CUDAModule& CUDAModule::operator=(CUDAModule&& other)
+{
+    std::swap(m_module, other.m_module);
```
Comment (dictoon) [inline src/appleseed/foundation/cuda/module.cpp]: `std::` qualifier redundant here.

**Pull request merged**